rust/frontend/src/internal_api.rs

```diff
@@ -0,0 +1,27 @@
+use std::cell::RefCell;
+use std::collections::HashMap;
+
+use Function;
+
+thread_local! {
+    pub(crate) static API: RefCell<HashMap<String, Function>> = RefCell::new(HashMap::new());
+}
+
+pub(crate) fn get(name: String) -> Option<Function> {
+    API.with(|hm| hm.borrow().get(&name).map(|f| f.clone()))
+}
+
+pub(crate) fn set(name: String, func: Function) {
+    API.with(|hm| {
+        (*hm.borrow_mut()).insert(name, func);
+    })
+}
+
+pub(crate) fn get_api(name: String) -> Function {
+    let mut func = get(name.clone());
+    if func.is_none() {
+        func = Function::get_function(&name, true, false);
+        set(name, func.clone().unwrap());
```

Comment:
are you sure these will never panic? if not, definitely propagate a `Result`

Reply:
Pretty sure this never panics as it's only used internally in `internal_api`and isn't pub.

Reply:
cool. can you add a comment or an `expect` which contains a message that this should never happen? [here's an example[(https://github.com/paritytech/parity-ethereum/blob/master/ethcore/src/executive.rs#L982). you can omit the "qed" if you decide to use an expect